src/meta/src/error.rs

```diff
@@ -0,0 +1,136 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::backtrace::Backtrace;
+use std::sync::Arc;
+
+use risingwave_pb::ProstFieldNotFound;
+use risingwave_rpc_client::error::RpcError;
+
+use crate::hummock::error::Error as HummockError;
+use crate::model::MetadataModelError;
+use crate::storage::MetaStoreError;
+
+pub type MetaResult<T> = std::result::Result<T, MetaError>;
+
+#[derive(thiserror::Error, Debug)]
+enum MetaErrorInner {
+    #[error("MetaStore transaction error: {0}")]
+    TransactionError(MetaStoreError),
+
+    #[error("MetadataModel error: {0}")]
+    MetadataModelError(MetadataModelError),
+
+    #[error("Hummock error: {0}")]
+    HummockError(HummockError),
+
+    #[error("Rpc error: {0}")]
+    RpcError(RpcError),
+
+    #[error("PermissionDenied: {0}")]
+    PermissionDenied(String),
+
+    #[error(transparent)]
+    Internal(anyhow::Error),
+}
+
+impl From<MetaErrorInner> for MetaError {
+    fn from(inner: MetaErrorInner) -> Self {
+        Self {
+            inner: Arc::new(inner),
+            backtrace: Arc::new(Backtrace::capture()),
+        }
+    }
+}
+
+#[derive(thiserror::Error, Clone)]
+#[error("{inner}")]
+pub struct MetaError {
+    inner: Arc<MetaErrorInner>,
+    backtrace: Arc<Backtrace>,
+}
+
+impl std::fmt::Debug for MetaError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        use std::error::Error;
+
+        write!(f, "{}", self.inner)?;
+        writeln!(f)?;
+        if let Some(backtrace) = self.inner.backtrace() {
+            write!(f, "  backtrace of inner error:\n{}", backtrace)?;
+        } else {
+            write!(f, "  backtrace of `MetaError`:\n{}", self.backtrace)?;
+        }
+        Ok(())
+    }
+}
+
+impl MetaError {
+    /// `MetaStore` transaction error.
+    pub fn transaction_error(e: MetaStoreError) -> Self {
+        MetaErrorInner::TransactionError(e).into()
+    }
+
+    /// Permission denied error.
+    pub fn permission_denied(s: String) -> Self {
+        MetaErrorInner::PermissionDenied(s).into()
+    }
+}
+
+impl From<MetadataModelError> for MetaError {
+    fn from(e: MetadataModelError) -> Self {
+        MetaErrorInner::MetadataModelError(e).into()
+    }
+}
+
+impl From<HummockError> for MetaError {
+    fn from(e: HummockError) -> Self {
+        MetaErrorInner::HummockError(e).into()
+    }
+}
+
+impl From<RpcError> for MetaError {
+    fn from(e: RpcError) -> Self {
+        MetaErrorInner::RpcError(e).into()
+    }
+}
+
+impl From<anyhow::Error> for MetaError {
+    fn from(a: anyhow::Error) -> Self {
+        MetaErrorInner::Internal(a).into()
+    }
+}
+
+impl From<MetaError> for tonic::Status {
+    fn from(err: MetaError) -> Self {
+        tonic::Status::internal(err.to_string())
```

Comment:
Different error should be casted into different tonic status. For example, `PermissionDenied` should be 
 `tonic::Status::permission_denied(err.to_string())` etc. I'm not sure if `invalid_argument`/`not_found` should have corresponding error type in `MetaError`, we can add those when meet in the future.